### PR TITLE
Java linked list task solution

### PR DESCRIPTION
**Pull request by mkazarian-1**

### CODE REVIEW

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 1]
```diff
@@ -3,45 +3,138 @@
...
+
+    private Node<T> searchNode(int index) {
+        checkIndex(index);
+        Node<T> currentNode = head;
+
+        int currentIndex = 0;
+        while (index != currentIndex) {
+            currentNode = currentNode.next;
+            currentIndex++;
+        }
```
Comment (sokimaaa) [inline src/main/java/core/basesyntax/MyLinkedList.java:34]: Don’t iterate your nodes from the head to tail or vice versa.

see common mistakes https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 2]
```diff
@@ -3,45 +3,138 @@
...
+
+    public MyLinkedList() {
+        head = new Node<>(null, null, null);
+        tail = head;
+        size = 0;
+    }
+
+    private void increaseSize(int value) {
+        size += value;
+    }
```
Comment (sokimaaa) [inline src/main/java/core/basesyntax/MyLinkedList.java:18]: Private methods and classes should be after public ones in your class

see common mistakes https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 3]
```diff
@@ -3,45 +3,138 @@
...
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int index = 0; index < size(); index++) {
+            if (currentNode.value == object
+                    || object != null && object.equals(currentNode.value)) {
+                remove(index);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
```
Comment (sokimaaa) [inline src/main/java/core/basesyntax/MyLinkedList.java:114]: with current implementation `remove(T object)` you iterates over list twice.
First, when you get the index.
Second, in the method `remove(int index)`, when you get the node. 

you can simplify it by adding `unlink(Node<T> node)` method.

Create a separate method`unlink(Node node)`.

See common mistakes https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html

**Review by sokimaaa — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 4]
```diff
@@ -3,45 +3,155 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+        size = 0;
+    }
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:12]:
```suggestion
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 5]
```diff
@@ -3,45 +3,155 @@
...
+        }
+        return currentNode;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(T value, Node<T> next, Node<T> prev) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:150]:
```suggestion
        public Node(Node<T> prev, T value, Node<T> next) {
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 6]
```diff
@@ -3,45 +3,155 @@
...
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unLinked(Node<T> node) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:94]:
```suggestion
    private void unlink(Node<T> node) {
```

**Review by Rommelua — APPROVED** [review]

**Pull request closed without merging**